### PR TITLE
Introduce commit for dao 5

### PR DESCRIPTION
**Pull request by gigbat**

### CODE REVIEW

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java [hunk 1]
```diff
@@ -1,46 +1,118 @@
...
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+            preparedStatement.setLong(1, id);
+
+            ResultSet resultSet = preparedStatement.executeQuery();
+            if (resultSet.next()) {
+                Manufacturer manufacturer = new Manufacturer();
+                manufacturer.setId(resultSet.getLong("manufacturer_id"));
+                manufacturer.setName(resultSet.getString("manufacturer_name"));
+                manufacturer.setCountry(resultSet.getString("manufacturer_country"));
```
Comment (yurwing) [inline src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]: realise it in a separate method

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java [hunk 2]
```diff
@@ -1,46 +1,118 @@
...
+                optionalManufacturer = Optional.of(manufacturer);
+            }
+        } catch (SQLException exception) {
+            throw new RuntimeException("Can't connect to DB", exception);
```
Comment (yurwing) [inline src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]: create your own exception
P.S it was in the checklist

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java [hunk 3]
```diff
@@ -1,46 +1,118 @@
...
+
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement preparedStatement = connection.prepareStatement(query,
+                        Statement.RETURN_GENERATED_KEYS)) {
+            preparedStatement.setString(1, manufacturer.getName());
+            preparedStatement.setString(2, manufacturer.getCountry());
+
+            preparedStatement.executeUpdate();
+            ResultSet generatedKeys = preparedStatement.getGeneratedKeys();
+            while (generatedKeys.next()) {
```
Comment (yurwing) [inline src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]:
```suggestion
            if (generatedKeys.next()) {
```

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java [hunk 4]
```diff
@@ -1,46 +1,118 @@
...
     @Override
     public List<Manufacturer> getAll() {
-        return Storage.getManufacturers();
+        List<Manufacturer> manufacturerList = new ArrayList<>();
+        String query = "SELECT * FROM manufacturers where manufacturer_deleted = false;";
+
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+            ResultSet resultSet = preparedStatement.executeQuery();
+            if (resultSet.next()) {
```
Comment (yurwing) [inline src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]:
```suggestion
            while (resultSet.next()) {
```

**Review by yurwing — CHANGES_REQUESTED** [review]

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java (hunk 1 above)
Comment (gigbat) [reply src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]: done

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java (hunk 2 above)
Comment (gigbat) [reply src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]: done, thanks

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java (hunk 3 above)
Comment (gigbat) [reply src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]: done

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java [hunk 5]
```diff
@@ -1,46 +1,118 @@
...
+        String query = "UPDATE manufacturers SET manufacturer_name = ?,"
+                + " manufacturer_country = ? where manufacturer_id = ?";
+
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+            preparedStatement.setString(1, manufacturer.getName());
+            preparedStatement.setString(2, manufacturer.getCountry());
+            preparedStatement.setLong(3, manufacturer.getId());
+
+            if (preparedStatement.executeUpdate() == 1) {
```
Comment (yurwing) [inline src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]:
```suggestion
            if (preparedStatement.executeUpdate() > 0) {
```

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java [hunk 6]
```diff
@@ -1,46 +1,118 @@
...
     @Override
     public boolean delete(Long id) {
-        return Storage.getManufacturers().removeIf(i -> i.getId().equals(id));
+        String query = "UPDATE manufacturers SET "
+                + "manufacturer_deleted = true where manufacturer_id = ?;";
+
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+            preparedStatement.setLong(1, id);
+            return preparedStatement.executeUpdate() == 1;
```
Comment (yurwing) [inline src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]:
```suggestion
            return preparedStatement.executeUpdate() > 0;
```

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java (hunk 4 above)
Comment (gigbat) [reply src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]: done

**Review by yurwing — CHANGES_REQUESTED** [review]

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java (hunk 5 above)
Comment (gigbat) [reply src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]: done

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java (hunk 6 above)
Comment (gigbat) [reply src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]: done

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java [hunk 7]
```diff
@@ -1,46 +1,123 @@
...
+
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+            ResultSet resultSet = preparedStatement.executeQuery();
+            while (resultSet.next()) {
+                Manufacturer manufacturer = new Manufacturer();
+                manufacturer.setId(resultSet.getLong("manufacturer_id"));
+                manufacturer.setName(resultSet.getString("manufacturer_name"));
+                manufacturer.setCountry(resultSet.getString("manufacturer_country"));
+                manufacturerList.add(manufacturer);
```
Comment (yurwing) [inline src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]: use method that you create for "get"

**Review by yurwing — CHANGES_REQUESTED** [review]

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java (hunk 7 above)
Comment (gigbat) [reply src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]: done

**Review by yurwing — APPROVED** [review]

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java [hunk 8]
```diff
@@ -1,46 +1,119 @@
...
+        String query = "INSERT INTO manufacturers (manufacturer_name, "
+                + "manufacturer_country) values (?, ?);";
+
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement preparedStatement = connection.prepareStatement(query,
+                        Statement.RETURN_GENERATED_KEYS)) {
+            preparedStatement.setString(1, manufacturer.getName());
+            preparedStatement.setString(2, manufacturer.getCountry());
+
+            preparedStatement.executeUpdate();
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]:
```suggestion
                + "manufacturer_country) values (?, ?);";
        try (Connection connection = ConnectionUtil.getConnection();
                PreparedStatement preparedStatement = connection.prepareStatement(query,
                        Statement.RETURN_GENERATED_KEYS)) {
            preparedStatement.setString(1, manufacturer.getName());
            preparedStatement.setString(2, manufacturer.getCountry());
            preparedStatement.executeUpdate();
```
these empty lines are redundant, they don't improve readability but make the code longer

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java [hunk 9]
```diff
@@ -1,46 +1,119 @@
...
+            ResultSet generatedKeys = preparedStatement.getGeneratedKeys();
+            if (generatedKeys.next()) {
+                manufacturer.setId(generatedKeys.getObject(1, Long.class));
+            }
+        } catch (SQLException exception) {
+            throw new NoConnectException("Can't connect to DB", exception);
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]: throw your custom exception here DataProcessingException (same with other methods)

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java [hunk 10]
```diff
@@ -1,46 +1,119 @@
...
+        return manufacturerList;
     }
 
     @Override
     public Manufacturer update(Manufacturer manufacturer) {
-        List<Manufacturer> dbManufacturer = Storage.getManufacturers();
-        for (int i = 0; i < dbManufacturer.size(); i++) {
-            if (manufacturer.getId().equals(dbManufacturer.get(i).getId())) {
-                dbManufacturer.set(i, manufacturer);
+        String query = "UPDATE manufacturers SET manufacturer_name = ?,"
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java:79]: let's not update deleted objects

File: src/main/java/internet/shop/exception/NoConnectException.java [hunk 11]
```diff
@@ -0,0 +1,9 @@
+package internet.shop.exception;
+
+import java.sql.SQLException;
+
+public class NoConnectException extends RuntimeException {
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/exception/NoConnectException.java]: you've put a checkmark that you reviewed your hw according to checklist, but here is still a mistake about exception from it, be more attentive next time 
image

File: src/main/java/internet/shop/service/impl/ManufacturerServiceImpl.java [hunk 12]
```diff
@@ -19,7 +19,8 @@ public Manufacturer create(Manufacturer manufacturer) {
 
     @Override
     public Manufacturer get(Long id) {
-        return manufacturerDao.get(id).get();
+        return manufacturerDao.get(id).orElseThrow(()
+                -> new RuntimeException("Data was deleted or not exist"));
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/service/impl/ManufacturerServiceImpl.java]:
```suggestion
                -> new RuntimeException("Manufacturer with id " + id + " was not found"));
```

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java (hunk 8 above)
Comment (gigbat) [reply src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]: thanks

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java (hunk 10 above)
Comment (gigbat) [reply src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java:79]: done

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java (hunk 9 above)
Comment (gigbat) [reply src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]: done

File: src/main/java/internet/shop/service/impl/ManufacturerServiceImpl.java (hunk 12 above)
Comment (gigbat) [reply src/main/java/internet/shop/service/impl/ManufacturerServiceImpl.java]: done

File: src/main/java/internet/shop/exception/NoConnectException.java (hunk 11 above)
Comment (gigbat) [reply src/main/java/internet/shop/exception/NoConnectException.java]: okey, done, i will more attentive

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java [hunk 13]
```diff
@@ -1,46 +1,119 @@
...
+        String query = "SELECT * FROM manufacturers where manufacturer_deleted = false;";
+
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement preparedStatement = connection.prepareStatement(query)) {
+            ResultSet resultSet = preparedStatement.executeQuery();
+            while (resultSet.next()) {
+                manufacturerList.add(setDataIntoManufacturer(resultSet));
+            }
+        } catch (SQLException exception) {
+            throw new RuntimeException("Can't get all data from DB", exception);
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]:
```suggestion
            throw new DataProcessingException("Can't get all data from DB", exception);
```
missed this one

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java [hunk 14]
```diff
@@ -1,46 +1,119 @@
...
+            preparedStatement.setString(2, manufacturer.getCountry());
+            preparedStatement.setLong(3, manufacturer.getId());
+
+            if (preparedStatement.executeUpdate() > 0) {
                 return manufacturer;
             }
+        } catch (SQLException exception) {
+            throw new DataProcessingException("Can't update the data - " + manufacturer, exception);
         }
         throw new RuntimeException("You can't update " + manufacturer + " in DB");
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]: and here

Comment (RomanDubovskyi) [inline src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]:
```suggestion
        throw new RuntimeException("No manufacturer " + manufacturer + " to update in DB");
```

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java [hunk 15]
```diff
@@ -1,46 +1,119 @@
...
 
     @Override
     public Manufacturer update(Manufacturer manufacturer) {
-        List<Manufacturer> dbManufacturer = Storage.getManufacturers();
-        for (int i = 0; i < dbManufacturer.size(); i++) {
-            if (manufacturer.getId().equals(dbManufacturer.get(i).getId())) {
-                dbManufacturer.set(i, manufacturer);
+        String query = "UPDATE manufacturers SET manufacturer_name = ?,"
+                + " manufacturer_country = ? where manufacturer_id = ? "
+                + "and manufacturer_deleted <> 1";
```
Comment (RomanDubovskyi) [inline src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]:
```suggestion
                + "and manufacturer_deleted = false;";
```

**Review by RomanDubovskyi — CHANGES_REQUESTED** [review]

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java (hunk 13 above)
Comment (gigbat) [reply src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]: exactly

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java (hunk 14 above)
Comment (gigbat) [reply src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]: did mire informative, thanks

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java (hunk 15 above)
Comment (gigbat) [reply src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]: ok, i remember

File: src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java (hunk 14 above)
Comment (gigbat) [reply src/main/java/internet/shop/dao/impl/ManufacturerDaoImpl.java]: done

**Review by RomanDubovskyi — APPROVED** [review]
good job

**Pull request merged**